core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/VariableReferenceGraph.java

```diff
@@ -0,0 +1,305 @@
+        graph.withNodeData(instanceList);
+        changed = new BitSet(instanceList.size());
+
+        graph.startBatchChange();
+        var visited = Collections.newSetFromMap(new IdentityHashMap<>());
+        for (var instance : instanceList) {
+            if (visited.add(instance.entity())) {
+                for (var variableId : variableReferenceToAfterProcessor.keySet()) {
+                    if (variableId.entityClass().isInstance(instance.entity())) {
+                        afterVariableChanged(variableId, instance.entity());
+                    }
+                }
+            }
+            entityToVariableReferenceMap.computeIfAbsent(instance.entity(), ignored -> new ArrayList<>())
+                    .add(instance);
+        }
+        for (var fixedEdgeEntry : fixedEdges.entrySet()) {
+            for (var toEdge : fixedEdgeEntry.getValue()) {
+                addEdge(fixedEdgeEntry.getKey(), toEdge);
+            }
+        }
+    }
+
+    public EntityVariableOrFactReference<?> lookup(VariableId variableReference, Object entity) {
+        return variableReferenceToInstanceMap.getOrDefault(variableReference, Collections.emptyMap()).get(entity);
+    }
+
+    public void addFixedEdge(EntityVariableOrFactReference<Solution_> from, EntityVariableOrFactReference<Solution_> to) {
+        if (from.id() == to.id()) {
+            return;
+        }
+        fixedEdges.computeIfAbsent(from, k -> new ArrayList<>()).add(to);
+    }
+
+    public void addEdge(EntityVariableOrFactReference<Solution_> from, EntityVariableOrFactReference<Solution_> to) {
+        if (from.id() == to.id()) {
+            return;
+        }
+        if (changed.isEmpty()) {
+            graph.startBatchChange();
+        }
+        var oldCount = counts[from.id()][to.id()]++;
+        if (oldCount == 0) {
+            graph.addEdge(from.id(), to.id());
+        }
+
+        markChanged(to);
+    }
+
+    public void removeEdge(EntityVariableOrFactReference<Solution_> from, EntityVariableOrFactReference<Solution_> to) {
+        if (from.id() == to.id()) {
+            return;
+        }
+        if (changed.isEmpty()) {
+            graph.startBatchChange();
+        }
+        var newCount = --counts[from.id()][to.id()];
+        if (newCount == 0) {
+            graph.removeEdge(from.id(), to.id());
+        }
+        markChanged(to);
+    }
+
+    public void markChanged(EntityVariableOrFactReference<?> node) {
+        if (changed.isEmpty()) {
+            graph.startBatchChange();
+        }
+        changed.set(node.id());
+    }
+
+    public void updateChanged() {
+        if (updating) {
+            // An exception occurred when processing a move
+            // Variable state is probably incorrect/will trigger a different
+            // exception, so return early.
+            return;
+        }
+
+        record AffectedEntity(Object entity, VariableUpdaterInfo variableUpdaterInfo) {
```

Comment:
The method is long enough already, even without data types being declared inside of it. Please move this type outside - you can still make it private.